File: 09-Smoke-Basin.js

const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin })

const lines = []
let maxRow
let maxCol

rl.on("line", data => {
  const line = data.split('').map(i=>Number.parseInt(i))
  lines.push(line)
  maxRow = lines.length - 1
  maxCol = lines[0].length - 1
})

rl.on('close', () => {
  console.log({maxRow, maxCol})
  console.log('partOne', partOne(lines))
  console.log('partTwo', partTwo(lines))
})

function cell(row,col) {
  return lines[row][col]
}

function getNeighbours(row, col) {
  const nums = []
  if (row > 0) {
    nums.push(cell(row-1, col))
  }
  if (row < maxRow) {
    nums.push(cell(row+1, col))
  }
  if (col > 0) {
    nums.push(cell(row, col-1))
  }
  if (col < maxCol) {
    nums.push(cell(row, col+1))
  }
  return nums
}

function isLow(row, col) {
  const current = cell(row,col)
  const neighbours = getNeighbours(row,col)

  return current < Math.min(...neighbours)
}

function partOne(lines) {
  let sum = 0
  for (let row = 0; row <= maxRow; row +=1 ) {
    for (let col = 0; col <= maxCol; col +=1 ) {
      if (isLow(row, col)) {
        sum+=cell(row, col) + 1
      }
    }
  }
  return sum
}

function partTwo(lines) {
  return 'todo'
}

File: 09-Smoke-Basin.hs

import Data.List.Split (splitOn)
import Utils (atoi,pp,flatten)

type Height = Int
type HeightMap = [[Height]]
type Coords = (Int,Int)
type Cell = (Coords,Height)

maxRow = 99
maxCol = 99

getneighbours :: HeightMap -> Coords -> [Height]
getneighbours heightmap (row, col) = map height neighbours
  where
    neighbours = filter withinMap [(row-1, col), (row+1, col), (row,  col-1), (row, col+1)]
    withinMap (r,c) = 0 <= r && 0 <= c && r <= maxRow && c <= maxCol
    height = getHeight heightmap

isLow heightmap (row, col) = current < minimum neighbours
   where
     current = getHeight heightmap (row, col)
     neighbours = getneighbours heightmap (row, col)

getHeight heightmap (r,c) = heightmap !! r !! c

toc xs n = map (\x -> (n,x)) xs
coords xs = map (toc [0..maxCol]) [0..maxRow]

solve1 input = sum lowpoints + length lowpoints
  where lowpoints = map (getHeight input) $ filter (isLow input) $ concat $ coords input

main = do
  contents <- getContents
  let heightmap = map (map atoi . tail . splitOn "") (lines contents)

  print $ solve1 heightmap -- 15   -- 535
  -- print $  input -- 1134 -- 1122700
